File: TakServer/lib/PTN.hs
module PTN where

import qualified Board as B
import Data.Char (digitToInt, isDigit, isLetter)
import Data.List (find)
import Data.Text (Text)
import qualified Data.Text as T

data PTN = PTN
  { site :: Maybe String
  , event :: Maybe String
  , date :: Maybe String
  , time :: Maybe String
  , p1 :: Maybe String
  , p2 :: Maybe String
  , clock :: Maybe String
  , ptnResult :: Maybe String
  , size :: Maybe Int
  , moves :: B.History
  } deriving (Show, Eq)

data PTNParseError
  = PTNParseError
  | PTNDirectionError
  | PTNCountError
  | PTNMoveError
  | PTNPositionError
  | PTNColorError
  | PTNStoneError
  | PTNSlideError
  deriving (Show, Eq)

parsePTN :: Text -> Either PTNParseError PTN
parsePTN input = do
  let lines' = T.lines input
  site' <- extractField "Site" lines'
  event' <- extractField "Event" lines'
  date' <- extractField "Date" lines'
  time' <- extractField "Time" lines'
  p1' <- extractField "Player1" lines'
  p2' <- extractField "Player2" lines'
  clock' <- extractField "Clock" lines'
  ptnResult' <- extractField "Result" lines'
  sizeStr' <- extractField "Size" lines'
  let size' =
        case sizeStr' of
          Just str -> read <$> Just str
          Nothing -> Nothing
  let moveLines = filter isMoveLine lines'
  moves' <-
    case parseMoves moveLines of
      Left e -> Left e
      Right [] -> Right []
      Right [x] -> Right [B.flipMoveColor x]
      Right (x:y:xs) -> Right (B.flipMoveColor x : B.flipMoveColor y : xs)
  return
    PTN
      { site = site'
      , event = event'
      , date = date'
      , time = time'
      , p1 = p1'
      , p2 = p2'
      , clock = clock'
      , ptnResult = ptnResult'
      , size = size'
      , moves = moves'
      }

isMoveLine :: Text -> Bool
isMoveLine line =
  let trimmed = T.strip line
   in not (T.null trimmed) && T.head trimmed `elem` ['0' .. '9']

parseMoves :: [Text] -> Either PTNParseError B.History
parseMoves ms = do
  movePairs <- traverse parseMovePair ms
  return (concat movePairs)

parseMovePair :: Text -> Either PTNParseError [B.Move]
parseMovePair movePair = do
  let moveStr = T.unpack movePair
  case words moveStr of
    [_, whiteMove] -> do
      whiteMove' <- parseSingleMove whiteMove B.White
      return [whiteMove']
    [_, whiteMove, blackMove] -> do
      whiteMove' <- parseSingleMove whiteMove B.White
      blackMove' <- parseSingleMove blackMove B.Black
      return [whiteMove', blackMove']
    _ -> Left PTNMoveError

parseSingleMove :: String -> B.Color -> Either PTNParseError B.Move
parseSingleMove moveStr color =
  let (moveStr', crush) =
        if last moveStr == '*'
          then (init moveStr, True)
          else (moveStr, False)
   in case moveStr' of
        [col, row]
          | isLetter col && isDigit row ->
            Right $
            B.PlaceFlat (B.Position (B.letterToCol col, digitToInt row), color)
        ['S', col, row]
          | isLetter col && isDigit row ->
            Right $
            B.PlaceStanding
              (B.Position (B.letterToCol col, digitToInt row), color)
        ['C', col, row]
          | isLetter col && isDigit row ->
            Right $
            B.PlaceCap (B.Position (B.letterToCol col, digitToInt row), color)
        _ -> parseSlideMove moveStr' color crush

parseSlideMove :: String -> B.Color -> Bool -> Either PTNParseError B.Move
parseSlideMove str color crush =
  case str of
    [countChar, col, row, dir] ->
      let pos = B.Position (B.letterToCol col, digitToInt row)
          dir' = charToDirection dir
          count = digitToInt countChar
       in case dir' of
            Right d -> Right $ B.Slide (pos, count, d, [count], color, crush)
            Left err -> Left err
    countChar:col:row:dir:dropsStr ->
      let pos = B.Position (B.letterToCol col, digitToInt row)
          dir' = charToDirection dir
          count = digitToInt countChar
          drops = map digitToInt dropsStr
       in case dir' of
            Right d -> Right $ B.Slide (pos, count, d, drops, color, crush)
            Left err -> Left err
    [col, row, dir] ->
      let pos = B.Position (B.letterToCol col, digitToInt row)
          dir' = charToDirection dir
       in case dir' of
            Right d -> Right $ B.Slide (pos, 1, d, [1], color, crush)
            Left err -> Left err
    _ -> Left PTNSlideError

charToDirection :: Char -> Either PTNParseError B.Direction
charToDirection c =
  case c of
    '+' -> Right B.Up
    '-' -> Right B.Down
    '>' -> Right B.Right
    '<' -> Right B.Left
    _ -> Left PTNDirectionError

extractField :: String -> [Text] -> Either PTNParseError (Maybe String)
extractField fieldName ls =
  case find (T.isPrefixOf (T.pack ("[" ++ fieldName ++ ": "))) ls of
    Just line ->
      Right $ Just $ T.unpack $ T.drop (length fieldName + 3) (T.init line)
    Nothing -> Right Nothing

directionToChar :: B.Direction -> Char
directionToChar B.Up = '+'
directionToChar B.Down = '-'
directionToChar B.Right = '>'
directionToChar B.Left = '<'

moveToText :: B.Move -> Text
moveToText (B.PlaceFlat ((B.Position (x, y)), color)) =
  T.pack $ B.colToLetter x : show y
moveToText (B.PlaceStanding ((B.Position (x, y)), color)) =
  T.pack $ "S" ++ [B.colToLetter x] ++ show y
moveToText (B.PlaceCap ((B.Position (x, y)), color)) =
  T.pack $ "C" ++ [B.colToLetter x] ++ show y
moveToText (B.Slide ((B.Position (x, y)), count, dir, drops, color, crush)) =
  T.pack $
  show count ++
  [B.colToLetter x] ++
  show y ++
  [directionToChar dir] ++
  concatMap show drops ++
  (if crush
     then "*"
     else "")

File: LambdaTak/src/ptn.c
#include "ptn.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

PTNParseError parsePTN(const char* input, PTN* ptn) {
    memset(ptn, 0, sizeof(PTN));

    char* lines[1024];
    int lineCount = 0;
    char* line = strtok((char*)input, "\n");
    while (line != NULL) {
        lines[lineCount++] = line;
        line = strtok(NULL, "\n");
    }

    for (int i = 0; i < lineCount; i++) {
        if (strncmp(lines[i], "[Site:", 6) == 0) {
            ptn->site = strdup(lines[i] + 6);
        } else if (strncmp(lines[i], "[Event:", 7) == 0) {
            ptn->event = strdup(lines[i] + 7);
        } else if (strncmp(lines[i], "[Date:", 6) == 0) {
            ptn->date = strdup(lines[i] + 6);
        } else if (strncmp(lines[i], "[Time:", 6) == 0) {
            ptn->time = strdup(lines[i] + 6);
        } else if (strncmp(lines[i], "[Player1:", 9) == 0) {
            ptn->player1 = strdup(lines[i] + 9);
        } else if (strncmp(lines[i], "[Player2:", 9) == 0) {
            ptn->player2 = strdup(lines[i] + 9);
        } else if (strncmp(lines[i], "[Clock:", 7) == 0) {
            ptn->clock = strdup(lines[i] + 7);
        } else if (strncmp(lines[i], "[Result:", 8) == 0) {
            ptn->result = strdup(lines[i] + 8);
        } else if (strncmp(lines[i], "[Size:", 6) == 0) {
            ptn->size = atoi(lines[i] + 6);
        }
    }

    ptn->moves = NULL;
    u8 flip = 0;
    for (int i = 0; i < lineCount; i++) {
        if (isdigit(lines[i][0])) {
            char* line_copy = strdup(lines[i]);
            if (!line_copy) return PTN_PARSE_ERROR;
            char* saveptr;
            strtok_r(line_copy, " ", &saveptr); // Skip move number
            Color current_color = WHITE;
            char* token;
            while ((token = strtok_r(NULL, " ", &saveptr)) != NULL) {
                Move move;
                PTNParseError err;
                if (flip < 2) {
                    err = parseMove(token, oppositeColor(current_color), &move);
                    flip++;
                }
                else {
                    err = parseMove(token, current_color, &move);
                }
                if (err != PTN_PARSE_OK) {
                    free(line_copy);
                    return err;
                }
                ptn->moves = addHistory(ptn->moves, move);
                current_color = (current_color == WHITE) ? BLACK : WHITE;
            }
            free(line_copy);
        }
    }

    return PTN_PARSE_OK;
}

PTNParseError parseMove(const char* moveStr, Color color, Move* move) {
    bool crush = (moveStr[strlen(moveStr) - 1] == '*');
    char* moveStrCopy = strdup(moveStr);
    if (crush) moveStrCopy[strlen(moveStrCopy) - 1] = '\0';

    if (strlen(moveStrCopy) == 2 && isalpha(moveStrCopy[0]) && isdigit(moveStrCopy[1])) {
        Position pos = {moveStrCopy[0] - 'a', moveStrCopy[1] - '1'};
        *move = createPlaceMove(pos, color, FLAT);
        free(moveStrCopy);
        return PTN_PARSE_OK;
    } else if (moveStrCopy[0] == 'S' && strlen(moveStrCopy) == 3 
            && isalpha(moveStrCopy[1]) && isdigit(moveStrCopy[2])) {
        Position pos = {moveStrCopy[1] - 'a', moveStrCopy[2] - '1'};
        *move = createPlaceMove(pos, color, STANDING);
        free(moveStrCopy);
        return PTN_PARSE_OK;
    } else if (moveStrCopy[0] == 'C' && strlen(moveStrCopy) == 3 
            && isalpha(moveStrCopy[1]) && isdigit(moveStrCopy[2])) {
        Position pos = {moveStrCopy[1] - 'a', moveStrCopy[2] - '1'};
        *move = createPlaceMove(pos, color, CAP);
        free(moveStrCopy);
        return PTN_PARSE_OK;
    } else {
        PTNParseError err = parseSlideMove(moveStrCopy, color, crush, move);
        free(moveStrCopy);
        return err;
    }
}

PTNParseError parseSlideMove(const char* str, Color color, bool crush, Move* move) {
    int count = 1;
    const char* ptr = str;

    if (isdigit(ptr[0])) {
        count = ptr[0] - '0';
        ptr++;
    }

    if (strlen(ptr) < 2) return PTN_POSITION_ERROR;
    Position pos = { ptr[0] - 'a', ptr[1] - '1' };
    ptr += 2;

    if (strlen(ptr) < 1) return PTN_DIRECTION_ERROR;
    Direction dir = charToDirection(ptr[0]);
    if (dir == -1) return PTN_DIRECTION_ERROR;
    ptr++;

    u8 drops[MAX_PICKUP];
    int dropCount = 0;

    while (*ptr != '\0' && isdigit(*ptr)) {
        printf("Drop: %c\n", *ptr);
        if (dropCount >= MAX_PICKUP) return PTN_COUNT_ERROR;
        drops[dropCount++] = *ptr - '0';
        ptr++;
    }

    if (dropCount == 0) {
        printf("No drops: count: %d\n", count);
        drops[0] = count;
        dropCount = 1;
    }

    for (;dropCount < MAX_PICKUP; dropCount++) {
        drops[dropCount] = 0;
    }

    // Create the slide move
    *move = createSlideMove(color, pos, dir, count, drops, crush ? CRUSH : NO_CRUSH);
    return PTN_PARSE_OK;
}

Direction charToDirection(char c) {
    switch (c) {
        case '+': return UP;
        case '-': return DOWN;
        case '>': return RIGHT;
        case '<': return LEFT;
        default: return -1;
    }
}

char directionToChar(Direction dir) {
    switch (dir) {
        case UP: return '+';
        case DOWN: return '-';
        case RIGHT: return '>';
        case LEFT: return '<';
        default: return '?';
    }
}

void freePTN(PTN* ptn) {
    if (ptn->site) free(ptn->site);
    if (ptn->event) free(ptn->event);
    if (ptn->date) free(ptn->date);
    if (ptn->time) free(ptn->time);
    if (ptn->player1) free(ptn->player1);
    if (ptn->player2) free(ptn->player2);
    if (ptn->clock) free(ptn->clock);
    if (ptn->result) free(ptn->result);
    if (ptn->moves) freeHistory(ptn->moves);
}
